File: android/gemcore/src/main/kotlin/com/gemwallet/android/ext/AddressFormatter.kt
package com.gemwallet.android.ext

import com.wallet.core.primitives.Chain
import uniffi.gemstone.GemAddressFormatStyle
import uniffi.gemstone.formatAddress

class AddressFormatter(
    private val address: String,
    private val chain: Chain? = null,
    private val style: Style = Style.Short,
) {
    sealed class Style {
        data object Short : Style()
        data object Full : Style()
        data class Extra(val count: Int) : Style()
    }

    fun value(): String {
        val gemstoneStyle = when (style) {
            is Style.Short -> GemAddressFormatStyle.Short
            is Style.Full -> GemAddressFormatStyle.Full
            is Style.Extra -> GemAddressFormatStyle.Extra(style.count.coerceAtLeast(0).toUInt())
        }
        return formatAddress(address, chain?.string, gemstoneStyle)
    }
}

File: ios/Packages/GemstonePrimitives/Sources/AddressFormatter.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Gemstone
import Primitives

public struct AddressFormatter {
    public enum Style {
        case short
        case full
        case extra(Int)
    }

    private let style: Self.Style
    private let address: String
    private let chain: Primitives.Chain?

    public init(
        style: Self.Style = .short,
        address: String,
        chain: Primitives.Chain?,
    ) {
        self.style = style
        self.address = address
        self.chain = chain
    }

    public func value() -> String {
        let gemstoneStyle: Gemstone.GemAddressFormatStyle = switch style {
        case .short: .short
        case .full: .full
        case let .extra(extra): .extra(extra: UInt32(clamping: max(extra, 0)))
        }

        return Gemstone.formatAddress(
            address: address,
            chain: chain?.rawValue,
            style: gemstoneStyle,
        )
    }
}
